### ISSUE TITLE
[BUG] Creating a CRTB or PRTB in Default namespace doesn't get deleted

### ISSUE THREAD

**Issue by JonCrowther** [rancher/rancher#42754]
**Rancher Server Setup**
- Rancher version: 2.8
- Installation option (Docker install/Helm Chart): Local

**Information about the Cluster**
- Kubernetes version: 1.26.7
- Cluster Type (Local/Downstream): Local

**User Information**
- What is the role of the user logged in? (Admin/Cluster Owner/Cluster Member/Project Owner/Project Member/Custom)
Admin

**Describe the bug**
When an admin creates a CRTB in the default namespace, it doesn't get deleted when the cluster is deleted.

**To Reproduce**
1. Create a downstream cluster
2. Create a Cluster Role
3. Create a CRTB without a namespace (ie Default)

```
apiVersion: management.cattle.io/v3
kind: ClusterRoleTemplateBinding
clusterName: <cluster ID>
metadata:
  name: <name>
roleTemplateName: <Cluster Role name>
userPrincipalName: <user principal name>
```
4. Delete the cluster

**Result**

The CRTB doesn't get deleted and fails to sync:
```
2023/09/11 18:45:19 [ERROR] error syncing 'default/dastest20202': handler mgmt-auth-crtb-controller: clusters.management.cattle.io "dastest1" not found, requeuing
2023/09/11 18:45:19 [ERROR] error syncing 'default/dastest202022': handler mgmt-auth-crtb-controller: clusters.management.cattle.io "dastest1" not found, requeuing
2023/09/11 18:45:19 [ERROR] error syncing 'default/dastestjonathan': handler mgmt-auth-crtb-controller: clusters.management.cattle.io "dastest1" not found, requeuing
2023/09/11 18:45:19 [ERROR] error syncing 'default/dastest1': handler mgmt-auth-crtb-controller: clusters.management.cattle.io "dastest1" not found, requeuing
```

**Expected Result**
Either don't allow CRTBs with namespaces that don't match the cluster name or clean them up properly

**Comment by JonCrowther** [comment]
Proposed solution:

1. Implement a webhook validation check that prevents users from creating CRTBs with mismatching `namespace` and `clusterName` fields.
2. If a user already has a CRTB with mismatched `namespace` and `clusterName`, allow updates to the CRTB but produce a warning informing the user that those fields are mismatched and that this is not supported behaviour

In a future release, deprecating `clusterName` will be investigated

**Comment by JonCrowther** [comment]
PRTBs have the same issue with mismatching `projectName` and `namespace`. They will also be addressed by this issue.

**Comment by JonCrowther** [comment]
In both PRTB and CRTBs, creating with a mismatched namespace and Cluster/Project Name is blocked now

**Comment by maxsokolovsky** [comment]
## QA Testing - PRTB

### Root cause

The `projectName` field on a PRTB must be specified as `clusterID:projectID`. The `projectID` part is supposed to match
`metadata.namespace`. Rancher never validates that. As a result, users are free to create PRTBs that cannot be deleted
when a project is deleted. Also, there are controller failures.

### What was fixed, or what changes have occurred

The webhook now ensures that the namespace and the projectID part of `projectName` must always be the same.

### Areas or cases that should be tested

### What areas could experience regressions?

### Steps

Use the following template for creating PRTBs:

```yaml
apiVersion: management.cattle.io/v3
kind: ProjectRoleTemplateBinding
metadata:
  name: {{name}}
  namespace: {{namespace}}
projectName: {{clusterID:projectID}}
roleTemplateName: configmaps-manage
userName: {{userName}}
```

#### Create a PRTB with a namespace that doesn't exist

1. Specify a `metadata.namespace` that doesn't exist and ensure you get a standard error from Kubernetes.

#### Create a PRTB with different namespace and project ID

1. Specify a valid `metadata.namespace`, one that matches the actual project namespace.
2. For `projectName`, specify a valid cluster ID but use any bad string for the project ID part. For
   example, `{{clusterID}}:blah`.
3. Ensure the **webhook** rejects your request.

**Comment by dasarinaidu** [comment]
I revalidated this issue on `v2.8-e5ce94a4c8572a8af176188172a57805e1dc76a4-head` & Webhook Version : `rancher/rancher-webhook:v0.4.0-rc9`

I did not see crtb related errors but I am seeing error messages related to project : `[project](mgmt-project-rbac-remove:)`

**Details:**

1. Provisioned rancher
2. Created downstream cluster
3. Created a project
4. Created PRTB and CRTB via kubectl
5. Deleted the project and check the error messages in logs
6. Created a project
7. Created PRTB and CRTB via kubectl
8. Delete cluster
9. Check for the error messages in logs

**Actual:**
1. Did not see any error messages related to crtb or prtb like earlier but seeing errors related to project as below.


--





```
2023/09/28 16:28:56 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:29:10 [ERROR] error syncing 'c-m-jltz97cf/p-sbdrz': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:29:26 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
W0928 16:29:39.595609      33 warnings.go:80] cluster.x-k8s.io/v1alpha3 Machine is deprecated; use cluster.x-k8s.io/v1beta1 Machine
2023/09/28 16:29:56 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:30:26 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing


2023/09/28 16:30:56 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:31:10 [ERROR] error syncing 'c-m-jltz97cf/p-sbdrz': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:31:26 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:31:56 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
W0928 16:32:24.674168      33 warnings.go:80] cluster.x-k8s.io/v1alpha3 Cluster is deprecated; use cluster.x-k8s.io/v1beta1 Cluster
2023/09/28 16:32:26 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:32:56 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:33:10 [ERROR] error syncing 'c-m-jltz97cf/p-sbdrz': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:33:26 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
W0928 16:33:36.414612      33 warnings.go:80] cluster.x-k8s.io/v1alpha3 MachineSet is deprecated; use cluster.x-k8s.io/v1beta1 MachineSet
W0928 16:33:47.426984      33 warnings.go:80] cluster.x-k8s.io/v1alpha3 MachineDeployment is deprecated; use cluster.x-k8s.io/v1beta1 MachineDeployment
2023/09/28 16:33:56 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:34:26 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:35:11 [ERROR] error syncing 'c-m-jltz97cf/p-sbdrz': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 16:36:26 [ERROR] error syncing 'c-m-47gk986g/p-9q79m': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
```

**Comment by dasarinaidu** [comment]
**"CATTLE_DEBUG": "true"
POD(Webhook) : Logs**
```
[ERROR] error syncing 'p-pcj48/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:12:03 [ERROR] error syncing 'p-dhb9z/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
```

**"CATTLE_DEBUG": "true"
Rancher Logs:**
```
2023/09/28 19:13:26 [ERROR] error syncing 'c-m-tpzhnwxn/p-cgvkz': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 19:13:26 [ERROR] error syncing 'c-m-tpzhnwxn/p-gbvsj': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 19:13:33 [ERROR] error syncing 'p-pcj48/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:13:33 [ERROR] error syncing 'p-dhb9z/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:13:34 [ERROR] error syncing 'p-dhb9z/prtb-mrz77': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
W0928 19:13:39.249334      32 warnings.go:80] Use tokens from the TokenRequest API or manually created secret-based tokens instead of auto-generated secret-based tokens.
2023/09/28 19:13:43 [ERROR] error syncing 'p-pcj48/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:13:43 [ERROR] error syncing 'p-dhb9z/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:13:44 [ERROR] error syncing 'p-dhb9z/prtb-mrz77': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:13:53 [ERROR] error syncing 'p-dhb9z/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:13:53 [ERROR] error syncing 'p-pcj48/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:13:54 [ERROR] error syncing 'p-dhb9z/prtb-mrz77': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
W0928 19:13:54.253697      32 warnings.go:80] Use tokens from the TokenRequest API or manually created secret-based tokens instead of auto-generated secret-based tokens.
2023/09/28 19:13:56 [ERROR] error syncing 'c-m-tpzhnwxn/p-cgvkz': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 19:13:56 [ERROR] error syncing 'c-m-tpzhnwxn/p-gbvsj': handler mgmt-project-rbac-remove: failed to remove finalizer on controller.cattle.io/mgmt-project-rbac-remove, requeuing
2023/09/28 19:14:03 [ERROR] error syncing 'p-pcj48/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:14:03 [ERROR] error syncing 'p-dhb9z/creator-project-owner': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
2023/09/28 19:14:04 [ERROR] error syncing 'p-dhb9z/prtb-mrz77': handler mgmt-auth-prtb-controller: clusters.management.cattle.io "c-m-bt572cc6" not found, requeuing
W0928 19:14:09.247961      32 warnings.go:80] Use tokens from the TokenRequest API or manually created secret-based tokens instead of auto-generated secret-based tokens.
```

**Comment by dasarinaidu** [comment]
I validated this issue on `v2.8-e5ce94a4c8572a8af176188172a57805e1dc76a4-head` & Webhook Version : `rancher/rancher-webhook:v0.4.0-rc9`. The below test cases passed and closing this ticket.

**Test Cases:**

1. Validate PRTB creation when Namespace is matching with Project Name(Valid and existed namespace)
2. Validate errors when Namespace is matching with Project Name(NOT existed namespace)
3. Validate errors when Namespace is NULL
4. Validate errors when Namespace is Not Valid
5. Validate errors when Namespace is Not matching with Project Name
6. Validate errors when Namespace is matching with Project Name

Note: I saw errors in log while deleting the cluster and reported a an issue. https://github.com/rancher/rancher/issues/42999
